### PR TITLE
BUG: Let check_exact_index default to True for integers

### LINKED ISSUE #57386: BUG: assert_index_equal defaults to check_exact_index=False on integers
Ref: https://github.com/pandas-dev/pandas/pull/57341#discussion_r1486812938

### PR DESCRIPTION
**Pull request by Aloqeely**
- [ ] closes #57386
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Review by mroeschke — COMMENTED** [review]
Sorry for the late review here.

1. Could you add to the docstring `exact` parameter that this applies to the `index` in 3.0?
2. Could you add a whatsnew note in `v3.0.0.rst`
3. Could you add some tests?

**Comment by Aloqeely** [comment]
No problem at all. Not sure how many tests are needed but I think this should be sufficient

File: pandas/_testing/asserters.py [hunk 1]
```diff
@@ -914,8 +920,23 @@ def assert_series_equal(
...
         )
+        left_index_dtypes = (
+            [left.index.dtype] if left.index.nlevels == 1 else left.index.dtypes
+        )
+        right_index_dtypes = (
+            [right.index.dtype] if right.index.nlevels == 1 else right.index.dtypes
+        )
+        check_exact_index = (
+            all(is_numeric_dtype(dtype) for dtype in left_index_dtypes)
+            and not any(is_float_dtype(dtype) for dtype in left_index_dtypes)
```
Comment (mroeschke) [inline pandas/_testing/asserters.py:931]: You could check `dtype.kind in "iu"` to check integer types in one go

File: pandas/_testing/asserters.py [hunk 2]
```diff
@@ -861,12 +861,19 @@ def assert_series_equal(
...
+        Index equivalence.
 
         .. versionchanged:: 2.2.0
 
             Defaults to True for integer dtypes if none of
             ``check_exact``, ``rtol`` and ``atol`` are specified.
+
+        .. versionchanged:: 3.0.0
+
+            When checking Index equivalence, the default value for check_exact
```
Comment (mroeschke) [inline pandas/_testing/asserters.py:874]: Could you specifically mention that integer dtypes check exact by default?

File: doc/source/whatsnew/v3.0.0.rst [hunk 3]
```diff
@@ -152,6 +152,7 @@ Other API changes
 - Updated :meth:`DataFrame.to_excel` so that the output spreadsheet has no styling. Custom styling can still be done using :meth:`Styler.to_excel` (:issue:`54154`)
 - pickle and HDF (``.h5``) files created with Python 2 are no longer explicitly supported (:issue:`57387`)
 - pickled objects from pandas version less than ``1.0.0`` are no longer supported (:issue:`57155`)
+- when comparing the indexes in :func:`testing.assert_series_equal`, the default value for check_exact is decided based on the :class:`Index` dtype, instead of the :class:`Series` dtype. (:issue:`57386`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v3.0.0.rst:155]: Same here about mentioning integer dtypes will check exact by default

File: pandas/_testing/asserters.py (hunk 1 above)
Comment (Aloqeely) [reply pandas/_testing/asserters.py:931]: Brilliant! That doesn't seem to work for multiindexes though as the dtype is always object (why?)

File: pandas/_testing/asserters.py (hunk 2 above)
Comment (Aloqeely) [reply pandas/_testing/asserters.py]: Of course!

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @Aloqeely

**Pull request merged**